### ISSUE TITLE
Flow analysis: allow assignments to promote variable types.

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#38725]
Currently flow analysis assumes that all assignments demote the type of the assigned variable to its declared type.  So for example, in the code:
```dart
f(int? i) {
  if (i == null) {
    i = 0; // (1)
  }
  print(i); // (2)
}
```
The assignment at (1) fails to promote the type of `i` from `int?` to `int`, and as a result, there is an error at (2).

**Comment by leafpetersen** [comment]
This seems to be fixed.  🎉

**Comment by stereotype441** [comment]
Thanks Leaf!  For the record this was fixed by 780af0fcfc557f739d59ebeca20550d5583ac1de.
